src/interface.jl

```diff
@@ -216,56 +227,75 @@ has_wings(duck::Duck)::Bool
 
     # generate code
     expr = quote
         function $func_name end
         BinaryTraits.register($can_type, $func_name,
-            ($(func_arg_types...),), $return_type)
+                ($(func_arg_types...),), tuple(($kwarg_names)...), $return_type)
     end
     display_expanded_code(expr)
     return esc(expr)
 end
 
 # Parsing function for @implement macro
 function parse_implement(can_type, by, sig)
-    usage = "Invalid @implement usage."
-    if !(can_type isa Symbol) || by !== :by || !(sig isa Expr)
-        throw(SyntaxError(usage))
-    end
+    usage = "usage: @implement <Type> by <function specification::<ReturnType>"
+    can_type isa Symbol && sig isa Expr && by === :by || throw(SyntaxError(usage))
 
     # Is return type specified?
-    has_return_type = sig.head == Symbol("::")
-    if has_return_type
+    if sig.head === Symbol("::") 
         return_type = sig.args[2]
         sig = sig.args[1]
     else
         return_type = :Any
     end
 
-    # TODO should we use @assert here?
-    @assert sig isa Expr
-    @assert sig.head === :call
+    sig isa Expr && sig.head === :call || throw(SyntaxError(usage))
 
     # parse signature
     func_name = sig.args[1]   # must be Symbol
     func_arg_names = Symbol[]
-    func_arg_types = Symbol[]
-    for (idx, x) in enumerate(sig.args[2:end])  # x must be Expr of 1 or 2 symbols
-        @assert x isa Expr
-        @assert x.head == Symbol("::")
-        @assert x.args |> length in [1,2]
-        if length(x.args) == 1
-            push!(func_arg_names, Symbol("x$idx"))
-            push!(func_arg_types, x.args[1])
-        else
-            push!(func_arg_names, x.args[1])
-            push!(func_arg_types, x.args[2])
+    func_arg_types = Any[]
+    func_kwarg_names = Symbol[]
+    firstarg = 2
+    if length(sig.args) >= 2 && sig.args[2] isa Expr && sig.args[2].head == :parameters
+        for x in sig.args[2].args
+            push!(func_kwarg_names, extract_name(x, nothing))
         end
+        firstarg += 1
     end
+    for (idx, x) in enumerate(sig.args[firstarg:end])  # x must be Expr of 1 or 2 symbols
+        push!(func_arg_names, extract_name(x, Symbol("x$idx")))
+        push!(func_arg_types, extract_type(x, :(Base.Bottom)))
+    end
+
+    return (func_name, func_arg_names, func_arg_types, func_kwarg_names, return_type)
+end
 
-    return (func_name, func_arg_names, func_arg_types, return_type)
+extract_name(x::Symbol, default) = x
+function extract_name(x::Expr, default)
+    n = length(x.args)
+    if x.head == Symbol("::")
+        n > 1 ? x.args[1] : default
+    elseif n >= 1
+        extract_name(x.args[1], default)
+    else
+        default
+    end
+end
+
+extract_type(::Symbol, default) = default
+function extract_type(x::Expr, default)
+    n = length(x.args)
+    if x.head == Symbol("::")
+        n > 1 ? x.args[2] : x.args[1]
+    elseif n >= 1
+        extract_type(x.args[1], default)
+    else
+        default
+    end
 end
```

Comment:
I like the idea of having these extract functions but somewhat intimidated by the number of conditional branches here.  I wonder if it would be clearer with this?  Also added some comments so that it's easier to maintain going forward.

```julia
# Extract name/type for a function argument without type specified
extract_name(x::Symbol, default) = x
extract_type(x::Symbol, default) = default

# Extract name/type for a function argument when it looks like `::T` or `x::T`
# For `::T`, it would be an Expr with 1 argument
# For `x::T`, it would be an Expr with 2 arguments

# Determine the "shape" of the function argument
argument_shape(x::Expr) = length(x.args) == 1 ? Val{:TypeOnly}() : Val{:NameAndType}()

extract_name(x::Expr, default) = extract_name(x, argument_shape(x), default)
extract_name(x::Expr, ::Val{:TypeOnly}, default) = default
extract_name(x::Expr, ::Val{:NameAndType}, default) = x.args[1]

extract_type(x::Expr, default) = extract_type(x, argument_shape(x), default)
extract_type(x::Expr, ::Val{:TypeOnly}, default) = x.args[1]
extract_type(x::Expr, ::Val{:NameAndType}, default) = x.args[2]
```

Reply:
I am not convinced.
In principle you can replace each `if .. elseif .. else ... end` construct with a sequence of separate method definitions, which dispatch on the condition values. (aside: `Val{xyz}() === Val(xyz)`).
If the conditions do not depend on compiler constants (as in our case), that would cause the compiler to  generate corresponding if-statements. So there is no performance advantage.
I prefer the programming style with the if-statements testing values, because it is easier to understand. For the argument types, a the other hand, I prefer the dispatch solution (to distinct the `x::Symbol` form the `x::Expr` case), because that gives me confidence on the fields and operations of `x` in side the method definition. 

Your approach is not complete, because it does not verify, if the expression has a `Symbol(::)`. It would be necessary to deliver default value (or better a syntax error), if that is not the case.
Introducing these missing conditions will make this approach even more convoluted.

Reply:
I agree that the multiple dispatch version doesn't provide additional performance benefit.  The only reason I came down that path is to find a way to write the code that's easier to interpret/understand. In fact, I do prefer if-elseif-else blocks as long as it's clear about the intention of each branch. Perhaps what's missing are comments around each conditional branch. 

One thing I hate about metaprogramming with AST is that the code is difficult to read.  All logic are going around `head` and `args` with length/specific index of the arg.  I can certainly take some time and understand what it does today but I will totally forget how it works in a month.

You have a good point below the recursive aspect of these functions, for which I didn't pay much attention to.  From a hindsight it does make sense to code it your way.  We can keep the if/else blocks but would you put some comments around the code? Imagine that someone other than yourself/me who may want to contribute in the future and that they may not have the years of experience like us ;-)

Reply:
done